File: src/main/java/arriba/fix/netty/FixMessageFrameDecoder.java
package arriba.fix.netty;

import java.util.Arrays;

import org.jboss.netty.buffer.ChannelBuffer;
import org.jboss.netty.buffer.ChannelBuffers;
import org.jboss.netty.channel.Channel;
import org.jboss.netty.channel.ChannelHandlerContext;
import org.jboss.netty.channel.ExceptionEvent;
import org.jboss.netty.handler.codec.frame.FrameDecoder;

import arriba.fix.Fields;
import arriba.fix.Tags;

public final class FixMessageFrameDecoder extends FrameDecoder {

    private static final byte[] CHECKSUM_BYTES = Tags.toByteArray(Tags.CHECKSUM);

    private byte nextFlagByte;
    private int nextFlagIndex;
    private boolean hasFoundFinalDelimiter;

    public FixMessageFrameDecoder() {
        this.reset();
    }

    @Override
    protected Object decode(final ChannelHandlerContext ctx, final Channel channel, final ChannelBuffer buffer) throws Exception {

        // TODO Should verify that tags are being received in correct order (e.g. checksume is last tag).
        // It is currently being assumed that tags are in the proper order.

        while ((this.nextFlagIndex = buffer.bytesBefore(this.nextFlagByte)) != -1) {
            // TODO Avoid reading unless looking for tag.
            final ChannelBuffer nextValueBuffer = buffer.readBytes(this.nextFlagIndex);
            buffer.readerIndex(buffer.readerIndex() + 1);

            if (Fields.EQUAL_SIGN == this.nextFlagByte) {
                this.nextFlagByte = Fields.DELIMITER;

                final byte[] tag = nextValueBuffer.array();
                if (Arrays.equals(CHECKSUM_BYTES, tag)) {
                    this.hasFoundFinalDelimiter = true;
                }
            } else if (Fields.DELIMITER == this.nextFlagByte) {
                this.nextFlagByte = Fields.EQUAL_SIGN;

                if (this.hasFoundFinalDelimiter) {
                    this.hasFoundFinalDelimiter = false;

                    this.reset();

                    final byte[] bufferBytes = buffer.array();
                    final byte[] fixMessageBytes = new byte[buffer.readerIndex()];
                    System.arraycopy(bufferBytes, 0, fixMessageBytes, 0, fixMessageBytes.length);

                    buffer.discardReadBytes();
                    return ChannelBuffers.copiedBuffer(fixMessageBytes);
                }
            }
        }

        return null;
    }

    @Override
    protected Object decodeLast(final ChannelHandlerContext ctx, final Channel channel, final ChannelBuffer buffer) throws Exception {
        // TODO Should probably be doing something here.
        System.out.println("decodelast called");
        return null;
    }

    @Override
    public void exceptionCaught(final ChannelHandlerContext ctx, final ExceptionEvent e) throws Exception {
        e.getCause().printStackTrace();
    }

    private void reset() {
        this.nextFlagIndex = -1;
        this.nextFlagByte = Fields.EQUAL_SIGN;
        this.hasFoundFinalDelimiter = false;
    }
}

File: src/main/scala/arriba/scala/fix/netty/FixMessageFrameDecoder.scala
package arriba.scala.fix.netty

import org.jboss.netty.handler.codec.frame.FrameDecoder
import com.google.common.collect.Lists
import org.jboss.netty.buffer.ChannelBuffer
import java.util.ArrayList
import org.jboss.netty.channel.{ExceptionEvent, Channel, ChannelHandlerContext}
import arriba.scala.fix.{SerializedField, Fields, Tags}

object FixMessageFrameDecoder {
  private val CHECKSUM_BYTES: Array[Byte] = Tags.toByteArray(Tags.CHECKSUM)
}

final class FixMessageFrameDecoder extends FrameDecoder {
  
  protected def decode(ctx: ChannelHandlerContext, channel: Channel, buffer: ChannelBuffer): AnyRef = {
    while ((({
      nextFlagIndex = buffer.bytesBefore(nextFlagByte); nextFlagIndex
    })) != -1) {
      val nextValueBuffer: ChannelBuffer = buffer.readBytes(nextFlagIndex)
      buffer.readerIndex(buffer.readerIndex + 1)
      if (Fields.EQUAL_SIGN == nextFlagByte) {
        tag = nextValueBuffer.array
        nextFlagByte = Fields.DELIMITER
        if (FixMessageFrameDecoder.CHECKSUM_BYTES eq tag) {
          hasFoundFinalDelimiter = true
        }
      }
      else if (Fields.DELIMITER == nextFlagByte) {
        value = nextValueBuffer.array
        nextFlagByte = Fields.EQUAL_SIGN
        //TODO fix this
        //serializedFields = (new SerializedField(tag, value)) :: serializedFields
        if (hasFoundFinalDelimiter) {
          hasFoundFinalDelimiter = false
          val tagsAndValuesCopy = new ArrayList[SerializedField](serializedFields)
          reset()
          tagsAndValuesCopy
        }
      }
    }
    null
  }

  protected override def decodeLast(ctx: ChannelHandlerContext, channel: Channel, buffer: ChannelBuffer): AnyRef = {
    println("decodelast called")
    null
  }

  override def exceptionCaught(ctx: ChannelHandlerContext, e: ExceptionEvent) {
    e.getCause.printStackTrace()
  }

  private def reset() {
    hasFoundFinalDelimiter = false
    tag = null
    value = null
    nextFlagIndex = -1
    nextFlagByte = Fields.EQUAL_SIGN
    hasFoundFinalDelimiter = false
    //serializedFields = Lists.newLinkedList
  }

  private var nextFlagByte: Byte = Fields.EQUAL_SIGN
  private var nextFlagIndex = -1
  private var tag: Array[Byte] = null
  private var value: Array[Byte] = null
  private var hasFoundFinalDelimiter: Boolean = false
  private var serializedFields = Lists.newLinkedList
}

